bridges/otellogr/logsink.go

```diff
@@ -0,0 +1,419 @@
+		opts = append(opts, log.WithSchemaURL(c.scope.SchemaURL))
+	}
+	return c.provider.Logger(c.scope.Name, opts...)
+}
+
+// Option configures a [LogSink].
+type Option interface {
+	apply(config) config
+}
+
+type optFunc func(config) config
+
+func (f optFunc) apply(c config) config { return f(c) }
+
+// WithInstrumentationScope returns an [Option] that configures the scope of
+// the [log.Logger] used by a [LogSink].
+//
+// By default if this Option is not provided, the LogSink will use a default
+// instrumentation scope describing this bridge package. It is recommended to
+// provide this so log data can be associated with its source package or
+// module.
+func WithInstrumentationScope(scope instrumentation.Scope) Option {
+	return optFunc(func(c config) config {
+		c.scope = scope
+		return c
+	})
+}
+
+// WithLoggerProvider returns an [Option] that configures [log.LoggerProvider]
+// used by a [LogSink] to create its [log.Logger].
+//
+// By default if this Option is not provided, the LogSink will use the global
+// LoggerProvider.
+func WithLoggerProvider(provider log.LoggerProvider) Option {
+	return optFunc(func(c config) config {
+		c.provider = provider
+		return c
+	})
+}
+
+// WithLevelSeverity returns an [Option] that configures the function used to
+// convert logr levels to OpenTelemetry log severities.
+//
+// By default if this Option is not provided, the LogSink will use a default
+// conversion function which adds an offset to the logr level to get the
+// OpenTelemetry severity. The offset is such that logr.Info("message")
+// converts to OpenTelemetry [log.SeverityInfo].
+func WithLevelSeverity(f func(int) log.Severity) Option {
+	return optFunc(func(c config) config {
+		c.levelSeverity = f
+		return c
+	})
+}
+
+// NewLogSink returns a new [LogSink] to be used as a [logr.LogSink].
+//
+// If [WithLoggerProvider] is not provided, the returned LogSink will use the
+// global LoggerProvider.
+func NewLogSink(options ...Option) *LogSink {
+	c := newConfig(options)
+	return &LogSink{
+		logger:        c.logger(),
+		levelSeverity: c.levelSeverity,
+	}
+}
+
+// LogSink is a [logr.LogSink] that sends all logging records it receives to
+// OpenTelemetry. See package documentation for how conversions are made.
+type LogSink struct {
+	// Ensure forward compatibility by explicitly making this not comparable.
+	noCmp [0]func() //nolint: unused  // This is indeed used.
+
+	name          string
+	logger        log.Logger
+	levelSeverity func(int) log.Severity
+	values        []log.KeyValue
+}
+
+// Compile-time check *Handler implements logr.LogSink.
+var _ logr.LogSink = (*LogSink)(nil)
+
+// log sends a log record to the OpenTelemetry logger.
+func (l *LogSink) log(err error, msg string, serverity log.Severity, kvList ...any) {
+	var record log.Record
+	record.SetTimestamp(time.Now())
+	record.SetBody(log.StringValue(msg))
+	record.SetSeverity(serverity)
+
+	if l.name != "" {
+		record.AddAttributes(log.String(nameKey, l.name))
+	}
```

Comment:
Instead of doing this the `WithName` method should create a new `l.logger`. The `nameKey` should not be needed.

https://github.com/open-telemetry/opentelemetry-go-contrib/pull/5357#discussion_r1577800199